cmd/coder/urls.go

```diff
@@ -14,13 +14,120 @@ import (
 )
 
 type urlsCmd struct{}
+type createURLCmd struct{}
+type delURLCmd struct{}
 
+// DevURL is the parsed json response record for a devURL from cemanager
 type DevURL struct {
+	ID     string `json:"id"`
 	URL    string `json:"url"`
 	Port   string `json:"port"`
 	Access string `json:"access"`
 }
 
+var urlAccessLevel = map[string]string{
+	"PRIVATE": "Only you can access",
+	"ORG":     "All members of your organization can access",
+	"AUTHED":  "Authenticated users can access",
+	"PUBLIC":  "Anyone on the internet can access this link",
+}
+
+func isAccessLevelValid(level string) bool {
+	_, ok := urlAccessLevel[level]
+	return ok
+}
+
+func (cmd createURLCmd) Spec() cli.CommandSpec {
+	return cli.CommandSpec{
+		Name:  "create",
+		Usage: "<env name> <port> <access>",
+		Desc: `create/update a devurl for external access
+  <access> is one of [PRIVATE | ORG | AUTHED | PUBLIC]`,
+	}
+}
+
+// Run() creates or updates a devURL, specified by env ID and port
+// (fl.Arg(0) and fl.Arg(1)), with access level (fl.Arg(2)) on
+// the cemanager.
+func (cmd createURLCmd) Run(fl *pflag.FlagSet) {
+	var envName = fl.Arg(0)
+	var port = fl.Arg(1)
+	var access = fl.Arg(2)
```

Comment:
We should provide a default value of private, and allow this to be optional.

We should also capitalize it for them, `public` or `org` should be valid input too.

Reply:
Sounds good

Reply:
I also wondered if I should do validation of the port arg here. Entering nonsensical value like 'fred' or '9999999' gives the server-side error, so do we care?

```$ ./coder create russ-env1 9999999 PRIVATE
Adding devurl for port 9999999
2020-06-25 08:39:37 ERROR	upsert devurl: http://localhost:8080/api/environments/5eecd3a0-c99dbfabfe3a9833e7172c5b/devurls
HTTP/1.1 400 Bad Request
Connection: close
Content-Length: 58
Coder-Version: 1.7.0-380-g7f5cbe16-ecf4366dc7
Content-Type: application/json
Date: Thu, 25 Jun 2020 15:39:20 GMT
Server: istio-envoy
Vary: Origin, Accept-Encoding
X-Envoy-Decorator-Operation: ingress-nginx.russtopia-rlmfuntoo.svc.cluster.local:80/*
X-Envoy-Upstream-Service-Time: 15
X-Powered-By: Express
```